### DISCUSSION TITLE
[Question][Hw3] Meaning of zext operation and bbenv

### DISCUSSION THREAD

**Discussion by Gimongjin** [prosyslab-classroom/cs424-program-reasoning#324] [Q&A]
Unfortunately I'm stuck before writing the actual code.

1. As I read in the llvm document, `zext` means convert the integer with a type (`i32, i64...`) to the another type by extending zero. 
Its usage was given like `%X = zext i32 257 to i64`. But in the example given in README.md, the zext operation was used as `%conv = zext %cmp`. I couldn't find any kind of example about using `zext` with this form. What is the meaning of `zext` in the example?

2. As given in `verifier.ml` and `extractor.ml`, when we run the `extract bb bbenv`, the `bbenv` is already containing the `'relation'` and `'free_var'` of every blocks in the `llm`. According to my understanding, `'relation'` (which type is `Z3.FuncDecl.func_decl`) means the predicate with the name of the block, which takes the free variables as the input and `Rel(free_vars)` means block named `Rel` can be entered with the values of free variables are in `free_vars`. 
I know that we have to build some rules by reading the each lines of instructions of the block and make it into a single `Z3 Expr.expr`, but I have no idea to how to combine the 'rules' and 'relations'. Is my understanding correct? I'm little confused since there are bunch of new types, structs and modules.

Thanks a lot for reading my question and giving any ideas.

**Comment by Hhro** [comment]
1. Sorry for the confusion. The correct instruction for the example is `zext i1 %cmp to i32`. As you said, it converts the integer value to another type by appending the 0 in front of the value. Since LLVM represents bool type with a one-length bit vector, where true is 1, and false is 0, it simply changes the type of boolean value without loss of meaning. 
  In the examples, `assume` gets the type `int` parameter in C. As you know, it is a 32-bit integer type. Therefore, LLVM converts the comparison result to the i32 value to meet the parameter type condition of the `assume`.

2. Basically, your understanding is correct. In this homework, a relation is the function declaration that maps free variables to the reachability of the block. And a rule is an implication that describes the reachability condition of the basic block.
  By composing the application of the relation (reachability condition of the current block) and the constraints derived from the instructions, you can make the reachability rule for the next block.
  After adding the whole rules to the fixed-point solver and querying whether the block which calls `is_assert_fail` is reachable, the solver will give you the answer.

Hope it helps.
